// @ionic/storage/__ivy_ngcc__/es2015/index.js
import { NgModule } from '@angular/core';
import { provideStorage, Storage, StorageConfigToken } from './storage';
import * as ɵngcc0 from '@angular/core';
export { Storage, StorageConfigToken };
export var IonicStorageModule = (function () {
    function IonicStorageModule() {
    }
    IonicStorageModule.forRoot = function (storageConfig) {
        if (storageConfig === void 0) { storageConfig = null; }
        return {
            ngModule: IonicStorageModule,
            providers: [
                { provide: StorageConfigToken, useValue: storageConfig },
                { provide: Storage, useFactory: provideStorage, deps: [StorageConfigToken] },
            ]
        };
    };
    /** @nocollapse */
    IonicStorageModule.ctorParameters = function () { return []; };
IonicStorageModule.ɵfac = function IonicStorageModule_Factory(t) { return new (t || IonicStorageModule)(); };
IonicStorageModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: IonicStorageModule });
IonicStorageModule.ɵinj = ɵngcc0.ɵɵdefineInjector({});
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(IonicStorageModule, [{
        type: NgModule,
        args: [{}]
    }], function () { return []; }, null); })();
    return IonicStorageModule;
}());
